llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp

```diff
@@ -905,8 +905,17 @@ Instruction *InstCombinerImpl::foldAddWithConstant(BinaryOperator &Add) {
 
   // (X | Op01C) + Op1C --> X + (Op01C + Op1C) iff the `or` is actually an `add`
   Constant *Op01C;
-  if (match(Op0, m_DisjointOr(m_Value(X), m_ImmConstant(Op01C))))
-    return BinaryOperator::CreateAdd(X, ConstantExpr::getAdd(Op01C, Op1C));
+  if (match(Op0, m_DisjointOr(m_Value(X), m_ImmConstant(Op01C)))) {
+    BinaryOperator *NewAdd =
+        BinaryOperator::CreateAdd(X, ConstantExpr::getAdd(Op01C, Op1C));
+    // Preserve the nsw flag iff the sum of Op01C and Op1C will not overflow
+    // so that there is a chance to make some other transformations.
+    // For some cases, sdiv can be converted to udiv when the newly created add
+    // carrying the nsw flag is one of its operands.
+    if (willNotOverflowSignedAdd(Op01C, Op1C, Add))
+      NewAdd->setHasNoSignedWrap(Add.hasNoSignedWrap());
```

Comment:
In the interest of completeness, can you please also preserve `nuw`? nuw can be preserved unconditionally, without an extra overflow check (see https://alive2.llvm.org/ce/z/YnJHnH).